### PR TITLE
build(gulpfile): Add custom transforms for rxjs imports

### PR DESCRIPTION
**Pull request by afragapane**
### Description of the Change
Move from gulp babel transforms to tsc custom transforms to transform rxjs imports.

### CODE REVIEW

**Comment by davemneo** [comment]
bor r-

File: gulpfile.js [hunk 1]
```diff
@@ -22,6 +23,115 @@ const rollup = require('rollup');
...
+
+  return statements;
+};
+
+const createCustomTransform = (context) => {
+  return (node) => {
+    const visitor = (node) => {
+      if (typescript.isSourceFile(node)) {
+        let statements = node.statements;
+        statements = transformImport(`rxjs`, transformRxjsImportText, statements);
```
Comment (dicarlo2) [inline gulpfile.js]: The way you're doing this is `O(n*m)` complexity, where `n` is the number of statements in the file and `m` is the number of imports to map. Please change it to `O(n)`. I'll leave it as an exercise to figure out how :)

Note, I fully realize it's probably not an issue here, but it's good practice to always use the most optimal solution.

File: gulpfile.js [hunk 2]
```diff
@@ -22,6 +23,115 @@ const rollup = require('rollup');
...
+          node.typeReferences,
+          node.hasNoDefaultLib,
+          node.libReferences,
+        );
+      }
+
+      return node;
+    };
+
+    typescript.visitEachChild(node, visitor, context);
```
Comment (dicarlo2) [inline gulpfile.js]: I don't think you need this.

File: gulpfile.js [hunk 3]
```diff
@@ -22,6 +23,115 @@ const rollup = require('rollup');
...
+            node.importClause,
+            typescript.createStringLiteral(`${moduleName}-browserify`),
+          );
+        }
+      }
+
+      return node;
+    };
+
+    typescript.visitEachChild(node, visitor, context);
```
Comment (dicarlo2) [inline gulpfile.js]: Ditto

File: gulpfile.js [hunk 4]
```diff
@@ -22,6 +23,115 @@ const rollup = require('rollup');
...
+    const visitor = (node) => {
+      if (typescript.isImportDeclaration(node)) {
+        const moduleName = node.moduleSpecifier.text;
+        if (isBuiltinModule(moduleName)) {
+          return typescript.updateImportDeclaration(
+            node,
+            node.decorators,
+            node.modifiers,
+            node.importClause,
+            typescript.createStringLiteral(`${moduleName}-browserify`),
```
Comment (dicarlo2) [inline gulpfile.js]: You need to check that all of these modules actually exist. (Hint: they don't, so we should only be transforming the ones that do)

File: tsconfig/tsconfig.es2017.browserify.cjs.json [hunk 5]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (dicarlo2) [inline tsconfig/tsconfig.es2017.browserify.cjs.json]: I don't think you need these browserify configs.

File: tsconfig.json [hunk 6]
```diff
@@ -34,6 +34,7 @@
     "noFallthroughCasesInSwitch": true,
     "forceConsistentCasingInFileNames": true
   },
+  "include": ["**/*"],
```
Comment (dicarlo2) [inline tsconfig.json]: Remove, shouldn't be necessary (this is the default)

File: packages/neo-one-worker/package.json [hunk 7]
```diff
@@ -8,5 +8,25 @@
     "rxjs": "^6.3.3",
     "tslib": "^1.9.3"
   },
-  "sideEffects": false
+  "sideEffects": false,
+  "browserify": {
```
Comment (dicarlo2) [inline packages/neo-one-worker/package.json]: Remove?

File: gulpfile.js [hunk 8]
```diff
@@ -22,6 +22,224 @@ const rollup = require('rollup');
...
+      return node;
+    };
+
+    typescript.visitEachChild(node, visitor, context);
+
+    return typescript.visitNode(node, visitor);
+  };
+};
+
+let packageBrowserifyMap = {};
```
Comment (dicarlo2) [inline gulpfile.js]: The gulp steps run in parallel, so you need to make sure the transform package.json always runs after the typescript transforms. Also need to verify that everything runs in the same process. Did you check that the output has the expected dependencies?

Comment (dicarlo2) [reply gulpfile.js]: Ah, I see that you did and it does run afterwards now. Cool.

**Comment by dicarlo2** [comment]
Looks like e2e tests are failing though.

**Pull request merged**